From: "Aneesh Kumar K.V" <aneesh.kumar@linux.vnet.ibm.com>
ext4: Don't allow nonextenst mount option for large filesystem

The block mapped inode format can address only blocks within
2**32. So with file system larger than 2**32 don't allow
noextents mount option.  There are multiple issues that need
to addressed in the long run.

a) When e2fsprogs support resizing an already existing  ext3
file system to greater than 2**32 we need to add support to block
allocator to handle growing already existing block  mapped inode
so that blocks allocated for them fall within 2**32

b) When we do that we should enable mounting  larger file system
with -o noextents mount option.

Till them we fail mount of large file systems with -o noextents.


Signed-off-by: Aneesh Kumar K.V <aneesh.kumar@linux.vnet.ibm.com>
Signed-off-by: Mingming Cao <cmm@us.ibm.com>
---
 fs/ext4/super.c |   15 +++++++++++++++
 1 file changed, 15 insertions(+)

Index: linux-2.6.26-rc8/fs/ext4/super.c
===================================================================
--- linux-2.6.26-rc8.orig/fs/ext4/super.c	2008-06-30 11:18:20.000000000 -0700
+++ linux-2.6.26-rc8/fs/ext4/super.c	2008-06-30 11:18:20.000000000 -0700
@@ -1004,6 +1004,7 @@ static int parse_options (char *options,
 	int qtype, qfmt;
 	char *qname;
 #endif
+	ext4_fsblk_t last_block;
 
 	if (!options)
 		return 1;
@@ -1326,6 +1327,20 @@ set_qf_format:
 			set_opt (sbi->s_mount_opt, EXTENTS);
 			break;
 		case Opt_noextents:
+			/*
+			 * When e2fsprogs support resizing an already existing
+			 * ext3 file system to greater than 2**32 we need to
+			 * add support to block allocator to handle growing
+			 * already existing block  mapped inode so that blocks
+			 * allocated for them fall within 2**32
+			 */
+			last_block = ext4_blocks_count(sbi->s_es) - 1;
+			if (last_block  > 0xffffffffULL) {
+				printk(KERN_ERR "EXT4-fs: Filesystem too "
+						"large to mount with "
+						"-o noextents options\n");
+				return 0;
+			}
 			clear_opt (sbi->s_mount_opt, EXTENTS);
 			break;
 		case Opt_i_version:
